example/advanced/server-io/advanced_server_io.cpp

```diff
@@ -0,0 +1,914 @@
+//
+// Copyright (c) 2016-2017 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+//------------------------------------------------------------------------------
+//
+// Example: Advanced server, one io_context per thread
+//
+//------------------------------------------------------------------------------
+
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+#include <boost/beast/websocket.hpp>
+#include <boost/beast/version.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/signal_set.hpp>
+#include <boost/asio/steady_timer.hpp>
+#include <boost/make_unique.hpp>
+#include <boost/bind.hpp>
+#include <boost/config.hpp>
+#include <algorithm>
+#include <cstdlib>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+
+using tcp = boost::asio::ip::tcp;               // from <boost/asio/ip/tcp.hpp>
+namespace http = boost::beast::http;            // from <boost/beast/http.hpp>
+namespace websocket = boost::beast::websocket;  // from <boost/beast/websocket.hpp>
+
+class round_robin
+{
+    using io_ptr = std::shared_ptr<boost::asio::io_context>;
```

Comment:
I prefer to just repeat the type, it is more clear

Reply:
We shouldn't need `shared_ptr` here it should be possible to simply use `vector<io_context>`.

Reply:
`vector` won't work because `io_context` is noncopyable.

Reply:
Ah yes you're right. `deque` works though.

Reply:
I'll try it out using a `deque`. Since it doesn't need to share ownership, a `vector<unique_ptr<io_context>>`  also works.

Reply:
That is ugly though, extra allocation and indirection per element.

Reply:
there's no need to resize, so I think `make_unique<io_context[]>(n)` is fine

Reply:
`make_unique` is not C++11

Reply:
How do you pass constructor arguments?

Reply:
There's no point of having more than 1 thread per `io_context` if you have load balancing across multiple `io_context`s., so a hardcoded concurrency hint of 1 is fine.

Reply:
Would there ever be a scenario where having several threads per multiple `io_context`'s would be beneficial? Say 2 `io_context`'s and 4 `thread`'s, 2 per context?

Reply:
I don't think so. And if there is, then it is such a niche use-case that I don't think it is worth addressing in this particular class.